### ISSUE TITLE
Text Spacing Bookmarklet test in Accessibility Insights

### ISSUE THREAD

**Issue by sahithikkss** [microsoft/accessibility-insights-web#3096]
#2564 
 Is your feature request related to a problem? Please describe.
Text Spacing defines the success criterion
https://www.w3.org/TR/WCAG21/#text-spacing
But tools are not yet standardized to test this area.

## Describe the desired outcome
A precise bookmarklet test which both users and testers can use to achieve/verify text spacing on a website.

## Describe alternatives you've considered
IFrames
https://www.html5accessibility.com/tests/tsbookmarklet.html
`javascript:(function(){ var style = document.createElement(%27style%27), styleContent = document.createTextNode(%27* { line-height: 1.5 !important; letter-spacing: 0.12em !important; word-spacing: 0.16em !important; } p{ margin-bottom: 2em !important; } %27); style.appendChild(styleContent ); document.getElementsByTagName(%27head%27)[0].appendChild(style); var iframes = document.querySelectorAll(%27iframe%27);for (var i=0; i<iframes.length; i++) {try{iframes[i].contentWindow.document.getElementsByTagName(%27head%27)[0].appendChild(style); } catch(error) { console.log(%27Bookmarklet error: %27+error)}}})();
`
CSS
https://dylanb.github.io/bookmarklets.html
`javascript:(function()%7B%0A%09document.querySelectorAll('*').forEach(function%20(el)%20%7B%0A%09%09if%20(el.nodeName.toLowerCase()%20%3D%3D%3D%20'p')%20%7B%0A%09%09%09el.style.setProperty('margin-bottom'%2C%20'2em'%2C%20'important')%3B%0A%09%09%7D%0A%09%09el.style.setProperty('line-height'%2C%20'1.5'%2C%20'important')%3B%0A%09%09el.style.setProperty('letter-spacing'%2C%20'0.12em'%2C%20'important')%3B%0A%09%09el.style.setProperty('word-spacing'%2C%20'0.16em'%2C%20'important')%3B%0A%09%7D)%0A%7D)()%3B`

Above bookmarklets does not cover all kinds of websites.

## Additional context
More info on bookmarklets is here
https://www.digitala11y.com/accessibility-bookmarklets-testing/

**Comment by jalkire** [comment]
Thanks for the feature request! We actually link to the first alternative you listed (https://www.html5accessibility.com/tests/tsbookmarklet.html) from the Text spacing requirement of the Text legibility test in [assessments](https://accessibilityinsights.io/docs/en/web/getstarted/assessment). Do you have an any specific issues with the bookmarklet or example pages where it fails?

**Comment by sahithikkss** [comment]
Please apply [bookmarklet 1 ]( https://www.html5accessibility.com/tests/tsbookmarklet.html)that works for iframes on this same page
https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview
This does not show any text spacing.

Now, on a different tab, open https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview
Apply [bookmarklet 2]( https://dylanb.github.io/bookmarklets.html) mentioned above.
Text spacing can be seen. 

Please let me know if this [page ](https://docs.microsoft.com/en-us/azure/azure-monitor/app/app-insights-overview) needs to be fixed or the test bookmarklet needs to be fixed.

**Comment by lisli1** [comment]
I applied these two bookmarklets to the App Insights Overview page you mentioned, and I am unable to repro the issue. I seemed to get the same text spacing behavior for both bookmarklets. 

For reference, I've tried it on Microsoft Edge Version 84.0.522.40 (Official build) (64-bit) and Google Chrome Version 84.0.4147.89 (Official Build) (64-bit). Perhaps it would help to know what browser you are running?

Gif of applying bookmarklet 1
![first-bookmarklet](https://user-images.githubusercontent.com/48259897/88435634-742c7000-cdb7-11ea-991a-9efca8021feb.gif)

Gif of applying bookmarklet 2
![second-bookmarklet](https://user-images.githubusercontent.com/48259897/88435647-78f12400-cdb7-11ea-8f48-4d2a08541065.gif)

**Comment by sahithikkss** [comment]
I used Microsoft Edge Version 84.0.522.40 (Official build) (64-bit) browser. But could not get text spacing as demonstrated above with the 1st bookmarklet. What could be the reason?

**Comment by karanbirsingh** [comment]
Copying over troubleshooting steps from an email thread last week with sahithikkss:

- When running the bookmarklet, check the browser console to see if there are any errors.
- Because the bookmarklet is Javascript that injects text-spacing styles into the page, you can edit it to output errors to the console if you don't see any from (1) or change it to something simple (i.e. console.log('hello world')) to see if it is an issue with the bookmarklet itself.
- Try using the bookmarklet on a different browser.

**Comment by sahithikkss** [comment]
I could see the same behavior (no text spacing with first script) on Firefox 79.0 (64-bit).
The only difference between the scripts is that one is not html-encoded and the other is html-encoded.
Plain scripts pose XSS vulnerability and could have been stopped from execution (by browser).
You may want to consider including the safer one between the two in the Accessibility Insights tool.

**Comment by karanbirsingh** [comment]
@sahithikkss do you mind trying these two bookmarklets to share whether they both create an alert box? https://codepen.io/karanbirsingh/pen/gOrbdYR The scripts above have more differences than HTML-encoding so this will help us narrow it down.

**Comment by sahithikkss** [comment]
Yes. Both did create alert box.

**Comment by LiLoDavis** [comment]
@sahithikkss Here's our proposed short-term solution. Would you please try out the instructions below and tell us what you think? 

For a longer-term solution, we've added feature request #3322 specifying a visual helper that applies the required styling to the root html document and to any html documents embedded in iframes.

### How to test
This procedure uses the browser Developer Tools (F12).
Note: This requirement does not apply to text in an image.

1. Apply styling to the root html document of the target page:
    1. With focus in the target page, open the Developer Tools by pressing F12 on your keyboard.
    1. In the Developer Tools, select the **Console** tab.
    1. Copy the following code, paste it into the Console, and press Enter on your keyboard.

`var style = document.createElement('style'), styleContent = document.createTextNode('* { line-height: 1.5 !important; letter-spacing: 0.12em !important; word-spacing: 0.16em !important; } p { margin-bottom: 2em !important; } iframe { border: thick dashed blue !important; background-color: yellow !important; }'); style.appendChild(styleContent ); document.getElementsByTagName('head')[0].appendChild(style);`

2. Examine the target page to observe the effect of the styling:
    1. Text in the page's root html document is expected to adopt these spacing properties:
        - **Letter spacing** (tracking) at 0.12 times the font size
        -  **Word spacing** at 0.16 times the font size
        - **Line height** (line spacing) at 1.5 times the font size
        - **Spacing after paragraphs** at 2 times the font size
    1. Any iframes in the page will be indicated by a dashed blue border and a yellow background. Text inside an iframe will not adopt styling applied to the root html document.
3. Apply styling to any iframe that contains visible text:
    1. Right click inside the iframe to open the context menu.
    1. In the context menu, select **Inspect**. The Developer Tools will switch to the Elements tab.
    1. In the Developer Tools, select the **Console** tab.
    1. Paste the styling code (provided in step 1c) into the console, then press Enter.
4. When you have applied styling to all parts of the page, verify that all text (including any text in iframes) adopts the expected spacing with no clipping or overlap.
5. Record your results:
    1. If you find any failures, select **Fail**, then add them as failure instances.
    1. Otherwise, select **Pass**.

**Comment by LiLoDavis** [comment]
@ferBonnin I presume we don't want this to be closed?

**Comment by sahithikkss** [comment]
Sorry for the delay. I tried the steps on a non-iframe negative scenario. Text Spacing is as expected.

But, it is still going to be manual test. Is my understanding correct? If so, the style being available as AccessibilityInsights**RecommendedBookMarklet**  (with a link on browser extension/desktop exe) would be better than debug-console steps as most UX developers and testers are already aware of using a bookmarklet.

**Comment by ferBonnin** [comment]
@sahithikkss yes, this will still be a manual test. 
we have filed a feature request for this #3322 too to make it assisted and we will be tracking users upvotes for it.

**Comment by ferBonnin** [comment]
this has been addressed with #3836
